types/wlr_keyboard_shortcuts_inhibit_v1.c

```diff
@@ -0,0 +1,212 @@
+
+	inhibitor->resource = wl_resource;
+	inhibitor->surface = surface;
+	inhibitor->seat_client = seat_client;
+	inhibitor->active = false;
+	wl_signal_init(&inhibitor->events.destroy);
+
+	inhibitor->surface_destroy.notify =
+		keyboard_shortcuts_inhibitor_handle_surface_destroy;
+	wl_signal_add(&surface->events.destroy, &inhibitor->surface_destroy);
+
+	inhibitor->seat_client_destroy.notify =
+		keyboard_shortcuts_inhibitor_handle_seat_client_destroy;
+	wl_signal_add(&seat_client->events.destroy, &inhibitor->seat_client_destroy);
+
+	wl_resource_set_implementation(wl_resource,
+		&keyboard_shortcuts_inhibitor_impl, inhibitor,
+		keyboard_shortcuts_inhibitor_v1_handle_resource_destroy);
+
+	wl_list_insert(&manager->inhibitors, &inhibitor->manager_link);
+	wlr_signal_emit_safe(&manager->events.new_inhibitor, inhibitor);
+}
+
+static void manager_handle_destroy(struct wl_client *client,
+		struct wl_resource *manager_resource) {
+	wl_resource_destroy(manager_resource);
+}
+
+static const struct zwp_keyboard_shortcuts_inhibit_manager_v1_interface
+keyboard_shortcuts_inhibit_impl = {
+	.destroy = manager_handle_destroy,
+	.inhibit_shortcuts = manager_handle_inhibit_shortcuts,
+};
+
+static void handle_display_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_keyboard_shortcuts_inhibit_manager_v1 *manager =
+		wl_container_of(listener, manager, display_destroy);
+	wlr_signal_emit_safe(&manager->events.destroy, manager);
+	wl_list_remove(&manager->display_destroy.link);
+	wl_global_destroy(manager->global);
+	free(manager);
+}
+
+static void keyboard_shortcuts_inhibit_bind(struct wl_client *wl_client,
+		void *data, uint32_t version, uint32_t id) {
+	struct wlr_keyboard_shortcuts_inhibit_manager_v1 *manager = data;
+
+	struct wl_resource *wl_resource = wl_resource_create(wl_client,
+		&zwp_keyboard_shortcuts_inhibit_manager_v1_interface,
+		version, id);
+	if (!wl_resource) {
+		wl_client_post_no_memory(wl_client);
+		return;
+	}
+
+	wl_resource_set_implementation(wl_resource,
+		&keyboard_shortcuts_inhibit_impl, manager, NULL);
+}
+
+struct wlr_keyboard_shortcuts_inhibit_manager_v1 *
+wlr_keyboard_shortcuts_inhibit_v1_create(struct wl_display *display) {
+	struct wlr_keyboard_shortcuts_inhibit_manager_v1 *manager =
+		calloc(1, sizeof(struct wlr_keyboard_shortcuts_inhibit_manager_v1));
+	if (!manager) {
+		return NULL;
+	}
+
+	wl_list_init(&manager->inhibitors);
+	wl_signal_init(&manager->events.new_inhibitor);
+	wl_signal_init(&manager->events.destroy);
+
+	manager->global = wl_global_create(display,
+		&zwp_keyboard_shortcuts_inhibit_manager_v1_interface, 1,
+		manager, keyboard_shortcuts_inhibit_bind);
+	if (!manager->global) {
+		free(manager);
+		return NULL;
+	}
+
+	manager->display_destroy.notify = handle_display_destroy;
+	wl_display_add_destroy_listener(display, &manager->display_destroy);
+
+	return manager;
+}
+
+void wlr_keyboard_shortcuts_inhibitor_v1_activate(
+		struct wlr_keyboard_shortcuts_inhibitor_v1 *inhibitor) {
```

Comment:
It would be nice to make this a no-op if the inhibitor is already active.

Reply:
Done.